### ISSUE TITLE
Support net5.0-windows10.0.17763.0 target for non-WinUI applications

### ISSUE THREAD

**Issue by mattico** [microsoft/appcenter-sdk-dotnet#1589]
### **Description**

Using AppCenter on .NET 5.0+ and Windows 10.0.17763.0+ causes frequent unhandled exceptions when AppCenter attempts to determine the app version.

### **Repro Steps**

Please list the steps used to reproduce your issue.

1. Change the TargetFramework of Microsoft.AppCenter.Test.WindowsDesktop.NetCore to `net5.0-windows10.0.17763.0`.
2. Run the tests.
3. The DeviceInformationHelperTests will fail.

### **Details**

Platform: Windows Desktop - .NET 6.0 - Windows 10.0.19041.0 - WPF
AppCenter SDK: 4.4.0
Windows 10.0.19043 x64

Stack Trace:

```
WinLaserlux7 Error: 0 : Unobserved Task Exception: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Exception of type 'System.InvalidOperationException' was thrown.)
 ---> System.InvalidOperationException: Exception of type 'System.InvalidOperationException' was thrown.
   at WinRT.ExceptionHelpers.ThrowExceptionForHR(Int32 hr)
   at ABI.Windows.ApplicationModel.IPackageStatics.global::Windows.ApplicationModel.IPackageStatics.get_Current()
   at Windows.ApplicationModel.Package.get_Current()
   at Microsoft.AppCenter.Utils.DeviceInformationHelper.GetAppVersion()
   at Microsoft.AppCenter.Utils.AbstractDeviceInformationHelper.GetDeviceInformation()
   at Microsoft.AppCenter.Utils.AbstractDeviceInformationHelper.<GetDeviceInformationAsync>b__9_0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__271_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.AppCenter.Channel.Channel.PrepareLogAsync(Log log, State state)
   at Microsoft.AppCenter.Channel.Channel.EnqueueAsync(Log log)
   --- End of inner exception stack trace ---
```

**Comment by AnastasiaKubova** [comment]
Hi! Thanks for getting in touch with us!
Does this issue happen only in tests? Time to time or always? Have you a chance to share a code snippet?

**Comment by AnastasiaKubova** [comment]
Hi! Thanks for getting in touch with us!
Does this issue happen only in tests? Time to time or always? Have you a chance to share a code snippet?
Also, could you please clarify do you use the App Center SDK in the WinUI project?

**Comment by mattico** [comment]
> Does this issue happen only in tests?

No, I found this issue while running my application. I just saw that the existing test could be used to reproduce the issue.

> Have you a chance to share a code snippet?

I could put one together tomorrow. Are you unable to reproduce with the tests? Is there something you're looking for? It would help me create it.

>  Also, could you please clarify do you use the App Center SDK in the WinUI project?

Yes. I'm using Microsoft.AppCenter.Analytics and Microsoft.AppCenter.Crashes in my WPF application.

**Comment by AnastasiaKubova** [comment]
I checked this behavior on a test app with the target version `net6.0-windows10.0.19041.0` and for me all work correctly. Could you please try to reproduce this issue on an empty app and let me know about the results?

**Comment by mattico** [comment]
This issue does prevent any log events from being reported. When TargetFramework is `net5.0-windows10.0.17763.0` I don't get any event showing up in Log Flow. Changing the TargetFramework to `net5.0-windows` causes events to show up in log flow again.

**Comment by AnastasiaKubova** [comment]
Thanks for sharing the details. Based on your project it doesn't look like a WinUI app. The target `net5.0-windows10.0.17763.0` should use only for WinUI applications. For other desktop apps, you should use `net5.0-windows` target.
Thus, the behavior you described [here](https://github.com/microsoft/appcenter-sdk-dotnet/issues/1589/#issuecomment-965569870) is expected.

**Comment by mattico** [comment]
I do not think that is correct. [The documentation recommends](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-enhance#net-5-and-later-use-the-target-framework-moniker-option) using a TFM with a windows version for calling WinRT APIs from Desktop Apps (WPF/WinForms). See [this example using WinRT APIs in a Windows Forms app](https://blogs.windows.com/windowsdeveloper/2020/09/03/calling-windows-apis-in-net5/). As explained in "[Windows Runtime APIs not supported in desktop apps](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api)" some WinRT APIs are only callable from UWP Apps, some APIs are only callable from MSIX Packaged Apps, but many APIs are callable from unpackaged Desktop Apps.

`Package.Current` [which is used by the AppCenterSDK and is causing this issue](https://github.com/microsoft/appcenter-sdk-dotnet/blob/0efb750d8a300b3095cf757244b0fe4e0851d87a/SDK/AppCenter/Microsoft.AppCenter.WindowsDesktop/Utils/DeviceInformationHelper.cs#L154-L159) is one such API which is only callable from MSIX Packaged Apps. Whether or not an app is MSIX packaged has nothing to do with what windows version it targets in the TFM.

**Comment by AnastasiaKubova** [comment]
Thanks for the details and your PR, I checked it.
Currently, the App Center SDK supports `net5.0-windows10.0.17763.0` target only for `WinUI in Desktop` and `WinUI in UWP` apps (how to create such apps please [see these docs](https://docs.microsoft.com/en-us/windows/apps/winui/winui3/release-notes/winui3-project-reunion-0.5-preview#install-winui-3---project-reunion-05-preview) and [changelog](https://github.com/microsoft/appcenter-sdk-dotnet/releases/tag/4.4.0)). The structure of your project is different from default WinUI applications.
Using `net5.0-windows10.0.17763.0` target in desktop applications can lead to the wrong behavior of App Center SDK, because for this target in App Center SDK is used other application lifecycle implementation.
I will label this issue as a feature request.

**Comment by mattico** [comment]
That's unfortunate. I can't find this limitation mentioned anywhere in the documentation, other than the changelog which doesn't mention that you *shouldn't* use `net5.0-windows10.0.17763.0` for non-WinUI apps. I'll create an issue for the docs.

**Comment by AnastasiaKubova** [comment]
Please note that the [changelog](https://github.com/microsoft/appcenter-sdk-dotnet/releases/tag/4.4.0) indicates the platform with this target that was added to App Center SDK support.

> Add support for platform **WinUI in Desktop**. The target version of **WinUI** apps should be **net5.0-windows10.0.17763.0** or higher.

Also, please pay attention that our docs for the `WPF/WinForms` applications [mention only platforms](https://docs.microsoft.com/en-us/appcenter/sdk/getting-started/wpf-winforms#1-prerequisites) which we currently support. This platform doesn't mention in docs for `WPF/WinForms` because currently the App Center SDK doesn't support this platform for these applications, only for UWP apps (please see [docs](https://docs.microsoft.com/en-us/appcenter/sdk/getting-started/uwp#1-prerequisites)).

**Comment by philipag** [comment]
I was able to get this working by granting identity to my non-packaged WPF desktop app as described here: [https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps](url)

This post has more detailed instructions: [https://nicksnettravels.builttoroam.com/sparse-package/](url)

**Comment by Lightczx** [comment]
Currently I just use 4.3.0 version of the SDK to avoid the WinUI compat behavior.
If there is any new release support WPF with RT, I will upgrade to that.

**Comment by nabeelio** [comment]
I'm getting this using `net5.0-windows10.0.18362.0`. I'm assuming it's related. Though this is a WPF application. Publishsinglefile is enabled. Should I open a new issue?

```
03-07 17:56:42|ERROR; UncaughtException::System.AggregateException
03-07 17:56:42|ERROR; UncaughtException:: (unwind) System.IO.FileNotFoundException
03-07 17:56:42|ERROR; UnhandledException StackTrace :    at System.Diagnostics.FileVersionInfo.GetVersionInfo(String fileName)
   at System.Windows.Forms.Application.GetAppFileVersionInfo()
   at System.Windows.Forms.Application.get_ProductVersion()
   at Microsoft.AppCenter.Utils.DeviceInformationHelper.GetAppVersion()
   at Microsoft.AppCenter.Utils.AbstractDeviceInformationHelper.GetDeviceInformation()
   at Microsoft.AppCenter.Utils.AbstractDeviceInformationHelper.<GetDeviceInformationAsync>b__9_0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__277_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.AppCenter.Channel.Channel.PrepareLogAsync(Log log, State state)
   at Microsoft.AppCenter.Channel.Channel.EnqueueAsync(Log log)
```

**Comment by nabeelio** [comment]
I just noticed it's in `System.Windows.Forms.Application` - I might have a config error.

**Comment by DmitriyKirakosyan** [comment]
Hi @nabeelio, yes this does looks like the same issue. The fix is already [merged](https://github.com/microsoft/appcenter-sdk-dotnet/pull/1615). It will be delivered in the upcoming release later this month.

**Comment by nabeelio** [comment]
@DmitriyKirakosyan Thank you. Any chance of having #1354 looked at? This was also related

**Comment by AnastasiaKubova** [comment]
Hi @nabeelio ! Could you please explain how this issue is related?

**Comment by nabeelio** [comment]
Hello, any updates on when 4.5.1 will be released?

**Comment by aleksandr-dorofeev** [comment]
The release will be ready in the next few days.

**Comment by nabeelio** [comment]
@aleksandr-dorofeev any updates? I'm about a month and half behind on my release schedule waiting on these fixes

**Comment by DmitriyKirakosyan** [comment]
Hi @nabeelio , as I replied in https://github.com/microsoft/appcenter-sdk-dotnet/issues/1620, it should be available later this week.
Thank you for your patience!

**Comment by nabeelio** [comment]
@DmitriyKirakosyan great, thank you so much

**Comment by DmitriyKirakosyan** [comment]
@nabeelio, the release 4.5.1 has been successfully released!
Closing the issue.
